### ISSUE TITLE
pressure level variables not validated by PrePARE due to wrong name?

### ISSUE THREAD

**Issue by oloapinivad** [PCMDI/cmor#409]
Hi all,

I have a few variables cmorized with data request 1.00.23  that fails the PrePARE validation stating that variable is not found in CMOR table.  We discussed it also in https://github.com/EC-Earth/ece2cmor3/issues/275
An example is zg_Emon, where the error involves zg (but I found this at least in ta_Emon and wap_6hrPlev too)

Such error actually makes sense, since the variable in the table Emon is zg27, not zg. 
````
        "zg27": {
            "frequency": "mon",
            "modeling_realm": "atmos",
            "standard_name": "geopotential_height",
            "units": "m",
            "cell_methods": "time: mean",
            "cell_measures": "area: areacella",
            "long_name": "Geopotential Height",
            "comment": "Geopotential is the sum of the specific gravitational potential energy relative to the geoid and the specific centripetal potential energy. Geopotential height is the geopotential divided by the standard acceleration due to gravity. It is numerically similar to the altitude (or geometric height) and not to the quantity with standard name height, which is relative to the surface.",
            "dimensions": "longitude latitude plev27 time",
            "out_name": "zg",
            "type": "real",
            "positive": "",
            "valid_min": "",
            "valid_max": "",
            "ok_min_mean_abs": "",
            "ok_max_mean_abs": ""
`````
While the ncdump -h | grep zg of my file reads: 

`````
$ ncdump -h zg_Emon_EC-Earth3P-HR_control-1950_r1i1p2f1_gr_195001-195012.nc  | grep zg
netcdf zg_Emon_EC-Earth3P-HR_control-1950_r1i1p2f1_gr_195001-195012 {
	float zg(time, plev, j, i) ;
		zg:standard_name = "geopotential_height" ;
		zg:long_name = "Geopotential Height" ;
		zg:comment = "Geopotential is the sum of the specific gravitational potential energy relative to the geoid and the specific centripetal potential energy. Geopotential height is the geopotential divided by the standard acceleration due to gravity. It is numerically similar to the altitude (or geometric height) and not to the quantity with standard name height, which is relative to the surface." ;
		zg:units = "m" ;
		zg:cell_methods = "time: mean" ;
		zg:cell_measures = "area: areacella" ;
		zg:history = "2018-09-24T11:56:03Z altered by CMOR: Reordered dimensions, original order: j i plev time." ;
		zg:missing_value = 1.e+20f ;
		zg:_FillValue = 1.e+20f ;
		zg:coordinates = "latitude longitude" ;
		:variable_id = "zg" ;
`````

However, I tried modifying manually with NCO the variable name to "zg" but I still get the same failed validation. Do you have any suggestion here?
Thanks a lot for any help,
Best
Paolo

**Comment by doutriaux1** [comment]
@oloapinivad could please paste a link to the failing file? It would help us greatly. Thanks.

**Comment by doutriaux1** [comment]
Got it. Thx!

**Comment by mauzey1** [comment]
@oloapinivad @doutriaux1 
I think I found the issue.  When you pass a file to PrePARE, it will get the "out_name" variable name and table name from the first two words of the file name.  It will then use the out_name and table name to look up the actual variable name in [LibCV/PrePARE/out_names_tests.json](https://github.com/PCMDI/cmor/blob/master/LibCV/PrePARE/out_names_tests.json).  It finds the right variable name by testing properties of the file such as whether or not it has 27 pressure levels.

Emon_zg, Emon_ta, and 6hrPlev_wap are missing from out_names_tests.json.  We could add zg27 and ta27 to the file but we would need to add a test in PrePARE to check for 4 pressure levels for wap4.

**Comment by taylor13** [comment]
Yes, I think you're correct that zg27 and ta27 for table Emon  need to be added to LibCV/PrePARE/out_names_tests.json. And wap4 needs to be added to 6hrPlev, *but* the extra coding checking the number of levels is only needed when another variable with outname wap is in the table.  In this case only wap4 appears, so we're o.k.

I think we should try to fix this as soon as we can so users can publish these fields successfully.

**Comment by taylor13** [comment]
Also,  don't we need to add to out_names_tests.json for table CFsubhr, how to treat "longitude", which will write an output file beginning "lon_CFsubhr_.....", and "latitude", which will write an output file beginning "lon_CFsubhr_....."?  In this case, these are not coordinate variables but actual variables, and their outnames differ from the table entry names.

Also, in https://github.com/PCMDI/cmor/issues/283#issuecomment-364202235 I counted 60 variables where the "entry" name and the outname differed.  Are we missing some variables in  LibCV/PrePARE/out_names_tests.json, which I think has only about 40 entries?

**Comment by mauzey1** [comment]
@taylor13 There are variables from tables CMIP6_cf3hr_0.json, CMIP6_cf3hr_1.json, and CMIP6_cf3hr_3.json that have out_names called "vertices_lat" and "vertices_lon".  These would be problematic for PrePARE since it will think the variable name is "vertices" and the table name is "lat" or "lon".  Are those tables still in use?  They have table dates from 2016.  Should they be removed from CMOR?

**Comment by taylor13** [comment]
Those tables (CMIP6_cf3hr_0.json, CMIP6_cf3hr_1.json, and CMIP6_cf3hr_3.json) are not needed in writing CMIP6 data (but perhaps they are needed for some of the test codes?).  The tables that should be included in the standard release are listed here: https://github.com/PCMDI/cmip6-cmor-tables/tree/master/Tables

**Comment by taylor13** [comment]
We need to check why ugrid has outname ugrido.
also why O18 and O17 seemed to be switched .
[Note: it's O18, O17, not 018, 017]

**Comment by mauzey1** [comment]
The issue with the naming of those variable outnames originates from the data request XML file.

http://proj.badc.rl.ac.uk/svn/exarch/CMIP6dreq/tags/01.00.28/dreqPy/docs/dreq.xml

Specifically, these lines:

```
<item defaultPriority="1" deflate="0" deflate_level="0" description="Water vapor path for water molecules that contain oxygen-18 (H2 18O)" frequency="mon" label="O18wv" mipTable="Emon" modeling_realm="atmos" mtid="MIPtable::Emon" positive="" prov="PMIP [PMIP-Amon-02]" provNote="" rowIndex="12" shuffle="0" stid="62eebe98-1617-11e7-a126-5404a60d96b5" title="O18 in water vapor" type="real" uid="6f68c8f2-9acb-11e6-b7ee-ac72891c3257" vid="590d4440-9e49-11e5-803c-0d0b866b59f3"/>
<item defaultPriority="1" deflate="0" deflate_level="0" description="Ratio of abundance of oxygen-17 (17O) atoms to oxgen-16 (16O) atoms in sea water" frequency="mon" label="sw18O" mipTable="Emon" modeling_realm="ocean" mtid="MIPtable::Emon" positive="" prov="PMIP [PMIP-Omon-02]" provNote="" rowIndex="0" shuffle="0" stid="f7e4000e-562c-11e6-a2a4-ac72891c3257" title="Isotopic Ratio of Oxygen-17 in Sea Water" type="real" uid="fdca5cc1-4d35-11e8-be0a-1c4d70487308" vid="fdca5cbf-4d35-11e8-be0a-1c4d70487308"/>
<item defaultPriority="1" deflate="" deflate_level="" description="Provide for models with unstructured grids only" frequency="fx" label="ugrid" mipTable="Ofx" modeling_realm="ocean" mtid="MIPtable::Ofx" positive="" prov="scanDreq.py" provNote="CMIP extra" rowIndex="0" shuffle="" stid="4f06575c-2674-11e7-96a5-ac72891c3257" title="UGRID Grid Information" type="real" uid="55e71213b0f6e3e098a0f0112beb2d50c2732f0a" vid="962e51dc-267b-11e7-96a5-ac72891c3257"/>
...
<item description="Ratio of abundance of oxygen-17 (17O) atoms to oxgen-16 (16O) atoms in sea water" label="sw17O" procComment="" procnote="" prov="PMIP [late]" provmip="PMIP" sn="isotope_ratio_of_17O_to_16O_in_sea_water_excluding_solutes_and_solids" title="Isotopic Ratio of Oxygen-17 in Sea Water" uid="fdca5cbf-4d35-11e8-be0a-1c4d70487308" unid="fd6fef50-3468-11e6-ba71-5404a60d96b5" units="1"/>
<item description="Water vapor path for water molecules that contain oxygen-18 (H2 18O)" label="sw18O" procComment="" procnote="" prov="CMIP6 endorsement [PMIP]" provmip="PMIP" sn="isotope_ratio_of_18O_to_16O_in_sea_water_excluding_solutes_and_solids" title="Mass of Water containing Oxygen-18 (H2 18O) in Layer" uid="590d4440-9e49-11e5-803c-0d0b866b59f3" unid="fd6fef50-3468-11e6-ba71-5404a60d96b5" units="1"/>
<item description="Ony required for models with unstructured grids: this label should be used for a file containing information about the grid structure, following the UGRID convention." label="ugrido" procComment="" procnote="" prov="CMIP6 Extra" provmip="CMIP6" sn="longitude" title="UGRID Grid Specification" uid="962e51dc-267b-11e7-96a5-ac72891c3257" unid="fd6e731e-3468-11e6-ba71-5404a60d96b5" units=""/>
```

**Comment by mauzey1** [comment]
@taylor13 What should we do about the O17, O18, and ugrid variable out_names?  I still have a pending pull request for adding the missing variables to out_names_tests.json for PrePARE.  Should I remove those variables from the current pull request until we figure out what to do with them?

**Comment by taylor13** [comment]
Yes, that's probably the right thing to do, but let's keep this issue open to remind us of the problem.  I have posted an issue here: https://github.com/cmip6dr/CMIP6_DataRequest_VariableDefinitions/issues/379 to try to get this fixed.

**Comment by taylor13** [comment]
Martin has promised to correct these problems in the next release of the data request.  Once we confirm, we can close.  [He hasn't done it yet.]

**Comment by taylor13** [comment]
These have been corrected in beta release of 01.00.30 of data request and propagated to CMOR tables, so I'm closing.

**Comment by mauzey1** [comment]
@taylor13 The 01.00.30 data request solves the out_names issue with sw17O, sw18O, and ugrid.  What about the other out_names not included in the current out_names_tests.json for PrePARE?

https://github.com/cmip6dr/CMIP6_DataRequest_VariableDefinitions/issues/379#issuecomment-458884568
According to this comment, the longitude and latitude out_names, lon and lat, are from CMIP5.  Should I remove those names from the changes in #412?

**Comment by taylor13** [comment]
@mauzey1 I have reviewed the CFsubhr variables "lon" and "lat".  They have CMOR table entry labels of "longitude" and "latitude", so in https://github.com/PCMDI/cmor/pull/412 we need to retain:
```
"CFsubhr_lat": {
    "has_variable_name": "latitude"
  },
  "CFsubhr_lon": {
    "has_variable_name": "longitude"
  },
```
That way when PrePARE finds a file with file name beginning "lat_CFsubhr ..." or "lon_CFsubhr", it will know to look for entries labeled "latitude" and "longitude" in the CMOR CFsubhr table.

The reference to CMIP5 is that in CMIP5 the table entry name was also "longitude" but the variable in the file was named "lon".  If there weren't that precedent, we probably would not have a different name for the entry and the variable in CMIP6.

By the way, I think it is a little confusing that the above out_names_test.json file says "has_variable_name" because the actual variable in the file is named "lat" or "lon", not latitude or longitude.  It would be less confusing if it said "has_table_entry_label" or "has_table_key" or "has_cmor_variable_name" or ???    Probably not worth changing though.

Unless I've got it all wrong (which is certainly possible), I think we can close this issue.
